fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java

```diff
@@ -548,41 +555,39 @@ public Response createObject(@HeaderParam(CONTENT_DISPOSITION) final ContentDisp
             return status(METHOD_NOT_ALLOWED).build();
         }
 
+        final var contentType = requestContentType != null ? requestContentType : DEFAULT_CONTENT_TYPE;
         // If request is an external binary, verify link header before proceeding
         final ExternalContent extContent = extContentHandlerFactory.createFromLinks(links);
-
-        final String contentType = extContent == null ? requestContentType.toString() : extContent.getContentType();
+        final String contentTypeStr = extContent == null ? contentType.toString() : extContent.getContentType();
 
         final String interactionModel = checkInteractionModel(links);
 
         final String fedoraId = identifierConverter().toInternalId(identifierConverter().toDomain(externalPath()));
-
         final String newFedoraId;
-        if (isBinary(interactionModel, requestContentType.toString(), requestContentType != null,
+
+        if (isBinary(interactionModel, contentType.toString(), requestContentType != null,
                 extContent != null)) {
             final Collection<String> checksums = parseDigestHeader(digest);
             final String originalFileName = contentDisposition != null ? contentDisposition.getFileName() : "";
-            newFedoraId = createResourceService.perform(transaction.getId(), getUserPrincipal(), fedoraId, slug, true,
-                    contentType, originalFileName, contentDisposition.getSize(), links, checksums, requestBodyStream,
-                    extContent);
+            newFedoraId = createResourceService.perform(transaction.getId(), getUserPrincipal(), fedoraId, slug,
+                    true, contentTypeStr,
+                    originalFileName, contentDisposition.getSize(), links, checksums, requestBodyStream, extContent);
         } else {
             final Model model = httpRdfService.bodyToInternalModel(externalPath(), requestBodyStream,
-                    requestContentType, identifierConverter());
-            newFedoraId = createResourceService.perform(transaction.getId(), getUserPrincipal(), fedoraId, slug, true,
-                    links, model);
+                    contentType, identifierConverter());
+            newFedoraId = createResourceService.perform(transaction.getId(), getUserPrincipal(), fedoraId, slug,
+                    true, links,
+                    model);
         }
-        // TODO: How to generate a response.
-        LOGGER.debug("Finished creating resource with path: {}", externalPath());
 
-        final FedoraResource resource;
+        LOGGER.debug("Finished creating resource: {}", newFedoraId);
+        transaction.commit();
```

Comment:
I see you ran into this commit order issue too. I'll just add that it should be `commitIfShortLived` so that it won't end a multi-transaction commit prematurely. That holds true for all the `commit` in this class actually.

Reply:
will fix.

Reply:
done.